### PR TITLE
Add org.mintguide.WebAppReader

### PR DESCRIPTION
**Pull request by ShvonderSiN**
WebAppReader is an application for  viewing and downloading websites offline on different platforms.

### Please confirm your submission meets all the criteria

- [X ] Please describe your application briefly.
- [X ] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I have [built][build] and tested the submission locally.
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
    - --env=QTWEBENGINEPROCESS_PATH=/app/bin/QtWebEngineProcess for proper works QtWebEngine
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
    Some libraries in binary form are necessary for wget 2 to work
- [X] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [X] The domain used for the application ID is controlled by the application developers either directly or through the code hosting (e.g. GitHub, GitLab, SourceForge, etc.). The [application id guidelines][app-id] are followed.
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission
[app-id]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id

### CODE REVIEW

File: org.mintguide.WebAppReader.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      - cp constants.py /app/
+      - cp settings.py /app/
+      - cp tools.py /app/
+      - cp tools.py /app/
+      - cp -r ui /app/
+      - cp -r pages /app/
+      - cp -r database /app/
+      - cp -r src /app/
+    sources:
+      - type: dir
```
Comment (bbhtt) [inline org.mintguide.WebAppReader.yaml:38]: Don't use `dir` and `path`, use git source or archive.

File: org.mintguide.WebAppReader.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        path: .
+        dest: app
+    post-install:
+      - install -Dm644 org.mintguide.WebAppReader.desktop /app/share/applications/org.mintguide.WebAppReader.desktop
+      - install -Dm644 src/logo.png /app/share/icons/hicolor/128x128/apps/org.mintguide.WebAppReader.png
+      - install -Dm755 wget2 /app/bin/wget2
+      - install -Dm755 libwget.so.2 /app/lib/libwget.so.2
```
Comment (bbhtt) [inline org.mintguide.WebAppReader.yaml:45]: It's in the runtime

File: org.mintguide.WebAppReader.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      - install -Dm755 libgnutls-dane.so.0 /app/lib/libgnutls-dane.so.0
+      - install -Dm755 libunbound.so.8 /app/lib/libunbound.so.8
+      - install -Dm755 libprotobuf-c.so.1 /app/lib/libprotobuf-c.so.1
+      - install -Dm755 libpython3.12.so.1.0 /app/lib/libpython3.12.so.1.0
```
Comment (bbhtt) [inline org.mintguide.WebAppReader.yaml:49]: You shouldn't copy paste libs like that, build what you need in the manifest, remove things that are in the runtime...

File: org.mintguide.WebAppReader.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  env:
+    QTWEBENGINE_DISABLE_SANDBOX: '1'
+modules:
+  - python3-requirements.json
+  - name: webappreader
+    subdir: app
+    buildsystem: simple
+    build-commands:
+      - cp webappreader.py /app/
+      - cp webappreader.sh /app/
```
Comment (bbhtt) [inline org.mintguide.WebAppReader.yaml:27]: It needs to be in `/app/bin` if this is the executable.

File: org.mintguide.WebAppReader.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
+app-id: org.mintguide.WebAppReader
+runtime: org.kde.Platform
+runtime-version: '6.6'
+sdk: org.kde.Sdk
+base: com.riverbankcomputing.PyQt.BaseApp
+base-version: '6.6'
+command: /app/webappreader.sh
```
Comment (bbhtt) [inline org.mintguide.WebAppReader.yaml:7]: Should not be a path

File: org.mintguide.WebAppReader.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+finish-args:
+  - --device=dri
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --filesystem=home
```
Comment (bbhtt) [inline org.mintguide.WebAppReader.yaml:12]: Why is this needed?

File: org.mintguide.WebAppReader.yaml [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+cleanup-commands:
+  - /app/cleanup-BaseApp.sh
+build-options:
+  env:
+    QTWEBENGINE_DISABLE_SANDBOX: '1'
+modules:
+  - python3-requirements.json
+  - name: webappreader
+    subdir: app
+    buildsystem: simple
```
Comment (bbhtt) [inline org.mintguide.WebAppReader.yaml:24]: Does this really have no buildsystem? Surely there is a better way instead of copying individual files.

File: org.mintguide.WebAppReader.yaml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      - cp settings.py /app/
+      - cp tools.py /app/
+      - cp tools.py /app/
+      - cp -r ui /app/
+      - cp -r pages /app/
+      - cp -r database /app/
+      - cp -r src /app/
+    sources:
+      - type: dir
+        path: .
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:39]: this is not gonna work.
`dir` for the source is always incorrect.

File: org.mintguide.WebAppReader.yaml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  - --filesystem=home
+  - --env=QTWEBENGINEPROCESS_PATH=/app/bin/QtWebEngineProcess
+  - --share=network
+cleanup-commands:
+  - /app/cleanup-BaseApp.sh
+build-options:
+  env:
+    QTWEBENGINE_DISABLE_SANDBOX: '1'
+modules:
+  - python3-requirements.json
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:21]: this is also missing.

File: org.mintguide.WebAppReader.yaml (hunk 2 above)
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:45]: in runtime, the wget2 version does not work correctly, there is a bug, so I added a newer one, in the form of a binary, it works normally

Comment (bbhtt) [reply org.mintguide.WebAppReader.yaml:45]: then build it here

File: org.mintguide.WebAppReader.yaml (hunk 3 above)
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:49]: these are dependencies for the newer version of wget2 to work

Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:49]: why does it require a different python?

Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:49]: it first sight you are just doing it wrong.

File: org.mintguide.WebAppReader.yaml (hunk 6 above)
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:12]: - --device=dri the program has a built-in browser, I hope that it works smoother thanks to this setting, if I'm wrong, I'll remove it.
- --filesystem=home this is for the possibility of saving pages to the hard disk

File: org.mintguide.WebAppReader.yaml (hunk 2 above)
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:45]: build from source?
So the program is accepted here only if the assembly in the manifest comes from the git repository? I can't add, I can't add binaries and libraries?

File: org.mintguide.WebAppReader.yaml (hunk 3 above)
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:49]: This is the only way my program started normally from runtime, I copied the libraries from fedora39

File: org.mintguide.WebAppReader.yaml (hunk 9 above)
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:21]: What you mean missing? Should I add all necessary files to this branch? And source code?

File: org.mintguide.WebAppReader.yaml (hunk 3 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:49]: don't do that.
build from source in the manifest.

File: org.mintguide.WebAppReader.yaml (hunk 9 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:21]: How do you think flatpak-builder will locate that file `python3-requirements.json` ? out of thin air?

File: org.mintguide.WebAppReader.yaml (hunk 5 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:7]: just to clarify: this it the **name** of the executable, that will be located from the `PATH`. Which is by default, `/app/bin`.

So:
1. it's wrong because you put a full path
2. it's not even pointing into a location in the `PATH`.

File: org.mintguide.WebAppReader.yaml (hunk 8 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:39]: And just to clarify, here `.` **currently** only contains the manifest and surely should not contain the upstream source.

File: org.mintguide.WebAppReader.yaml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      - cp tools.py /app/
+      - cp -r ui /app/
+      - cp -r pages /app/
+      - cp -r database /app/
+      - cp -r src /app/
+    sources:
+      - type: dir
+        path: .
+        dest: app
+    post-install:
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:41]: Whichever non functional state the build is right now, I can say that I don't see the metainfo (appstream) file being installed. It is mandatory.

**Comment by ShvonderSiN** [comment]
Changes have been made, please re-consider

File: org.mintguide.WebAppReader.yaml [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+      - install -D /dev/null /app/share/applications/org.mintguide.WebAppReader.desktop
+      - |
+        echo '[Desktop Entry]' > /app/share/applications/org.mintguide.WebAppReader.desktop
+        echo 'Name=WebAppReader' >> /app/share/applications/org.mintguide.WebAppReader.desktop
+        echo 'Comment=Offline web content reader' >> /app/share/applications/org.mintguide.WebAppReader.desktop
+        echo 'Exec=/app/bin/webappreader.sh %U' >> /app/share/applications/org.mintguide.WebAppReader.desktop
+        echo 'Icon=org.mintguide.WebAppReader' >> /app/share/applications/org.mintguide.WebAppReader.desktop
+        echo 'Terminal=false' >> /app/share/applications/org.mintguide.WebAppReader.desktop
+        echo 'Type=Application' >> /app/share/applications/org.mintguide.WebAppReader.desktop
+        echo 'Categories=Utility;' >> /app/share/applications/org.mintguide.WebAppReader.desktop
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:44]: don't do that. just modify the desktop file in place with `desktop-file-edit`.
don't generate a file with echo.
don't install /dev/null.

File: org.mintguide.WebAppReader.yaml [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+      - 'echo "#!/bin/sh" > /app/bin/webappreader.sh'
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:45]: this should be source of `type: script`.

File: org.mintguide.WebAppReader.yaml (hunk 10 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:60]: this hasn't be solved. it's mandatory.

File: org.mintguide.WebAppReader.yaml [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+      - 'echo "python3 /app/webappreader.py \"$@\"" >> /app/bin/webappreader.sh'
+      - chmod +x /app/bin/webappreader.sh
+      - cp -rT /run/build/webappreader/database /app/database
+      - cp -rT /run/build/webappreader/pages /app/pages
+      - cp -rT /run/build/webappreader/src /app/src
+      - cp -rT /run/build/webappreader/ui /app/ui
+      - cp /run/build/webappreader/constants.py /app
+      - cp /run/build/webappreader/settings.py /app
+      - cp /run/build/webappreader/tools.py /app
+      - cp /run/build/webappreader/webappreader.py /app
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:55]:
```suggestion
      - cp -rT database /app/database
      - cp -rT pages /app/pages
      - cp -rT src /app/src
      - cp -rT ui /app/ui
      - install -Dm644 constants.py -t /app
      - install -Dm644 settings.py -t /app
      - install -Dm644 tools.py -t /app
      - install -Dm644 webappreader.py -t /app
```

note that installing directly in `/app` doesn't look right.

File: org.mintguide.WebAppReader.yaml [hunk 14]
```diff
@@ -0,0 +1,62 @@
+app-id: org.mintguide.WebAppReader
+runtime: org.kde.Platform
+runtime-version: '6.6'
+sdk: org.kde.Sdk
+base: com.riverbankcomputing.PyQt.BaseApp
+base-version: '6.6'
+command: /app/bin/webappreader.sh
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:7]:
```suggestion
command: webappreader.sh
```

as previously said this should NOT be a full path.

**Comment by ShvonderSiN** [comment]
Changes have been made, please re-consider

**Comment by ShvonderSiN** [comment]
Changes have been made, please re-consider. Also added metainfo file

File: org.mintguide.WebAppReader.yaml [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    sources:
+      - type: git
+        url: https://gitverse.ru/sc/shekin/webappreader.git
+        commit: 88d98aa310103be2704a3908a0cec1aec65e8a01
+      - type: file
+        path: org.mintguide.WebAppReader.desktop
+      - type: script
+        commands:
+          - 'python3 /app/webappreader/webappreader.py "$@"'
+        dest-filename: webappreader.sh
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:53]: this is never installed.

Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:53]: ah nevermind. 

There is no reason to have a separate `post-install` which is why I thought it wasn't install.

File: org.mintguide.WebAppReader.yaml (hunk 10 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:56]: `post-install` should be merged `build-commands` since it's `simple` build system.

File: org.mintguide.WebAppReader.yaml (hunk 15 above)
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:53]: I tried few times... Successfully built and installed... What wrong exactly?

File: org.mintguide.WebAppReader.yaml [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+      - cp -r pages /app/webappreader/pages
+      - cp -r src /app/webappreader/src
+      - cp -r ui /app/webappreader/ui
+      - install -Dm644 constants.py -t /app/webappreader
+      - install -Dm644 settings.py -t /app/webappreader
+      - install -Dm644 tools.py -t /app/webappreader
+      - install -Dm644 webappreader.py -t /app/webappreader
+    sources:
+      - type: git
+        url: https://gitverse.ru/sc/shekin/webappreader.git
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:46]: this repositry is 404. So it won't work.

File: org.mintguide.WebAppReader.yaml [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+        commit: 88d98aa310103be2704a3908a0cec1aec65e8a01
+      - type: file
+        path: org.mintguide.WebAppReader.desktop
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:49]: this file doesn't exist.

Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:49]: ohh))) yes.. not anouth rights)) i will move it to another hosting//.. thx

**Comment by ShvonderSiN** [comment]
Changes have been made, please re-consider. Moved repository to github

File: org.mintguide.WebAppReader.yaml (hunk 17 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:49]: this file still doesn't exist.

File: org.mintguide.WebAppReader.yaml (hunk 10 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:56]: this wasn't a suggestion.

**Comment by ShvonderSiN** [comment]
Changes have been made, please re-consider. added missing file

File: org.mintguide.WebAppReader.yaml (hunk 10 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:56]: this still not done.

File: org.mintguide.WebAppReader.yaml [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    buildsystem: simple
+    build-commands:
+      - ./bootstrap
+      - ./configure --prefix=/app
+      - make
+      - make install
+    sources:
+      - type: git
+        url: https://gitlab.com/gnuwget/wget2.git
+        commit: 657e3d22c488a2d1f173e208e253995e15f8b6db
```
[the remark this answers is not in the document]
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:30]: Yes, it is installed without problems, version 2.1.... The version 2.0 built into runtime does not work well... that's why I'm building version 2.1.
here
```bash
  - name: wget2 #  proper workable version = 2.1
    buildsystem: simple
    build-commands:
      - ./bootstrap
      - ./configure --prefix=/app
      - make
      - make install
    sources:
      - type: git
        url: https://gitlab.com/gnuwget/wget2.git
        commit: 657e3d22c488a2d1f173e208e253995e15f8b6db
```

[the remark this answers is not in the document]
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:30]: just apply the suggestion.

File: org.mintguide.WebAppReader.yaml [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+sdk: org.kde.Sdk
+base: com.riverbankcomputing.PyQt.BaseApp
+base-version: '6.6'
+command: webappreader.sh
+finish-args:
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --filesystem=home  # to be able to save html pages to a hard drive
+  - --share=network # it is necessary to download html from the Internet
+  #- --device=dri # For the Qt webengine module to work perfectly with gpu
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:13]: why is this commented out?

Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:13]: I'm still testing, but when it's commented out, there are such errors in the terminal.... But the program works flawlessly...
```bash
test@fedora:~/Documents$ flatpak run org.mintguide.WebAppReader
qt.qpa.qgnomeplatform: Could not find color scheme  ""
libEGL warning: wayland-egl: could not open /dev/dri/renderD128 (No such file or directory)
MESA: error: ZINK: failed to choose pdev
libEGL warning: egl: failed to create dri2 screen
[3:43:0322/120707.909037:ERROR:vaapi_wrapper.cc(848)] Could not get a valid VA display
[51:9:0322/120708.116394:ERROR:command_buffer_proxy_impl.cc(128)] ContextResult::kTransientFailure: Failed to send GpuControl.CreateCommandBuffer.
js: TypeError: Failed to fetch
Fontconfig error: Cannot load default config file: No such file: (null)
js: crbug/1173575, non-JS module files deprecated.
js: TypeError: Failed to fetch
```
Here guys tell me that I do not need to use additional options if there is no need, so I left it commented out

Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:13]: and when it is not commented out?

Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:13]: If you'll excuse me, then I would rather turn it on, since the browser will probably work more smoothly.

Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:13]: which is why I was asking why it was commented out.

File: org.mintguide.WebAppReader.yaml (hunk 6 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml]: Just to clarify `--device=dri` was NEVER the question here.
Not sure how it look so ambiguous:

image

File: org.mintguide.WebAppReader.yaml (hunk 19 above)
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:13]: Because I was afraid that you wouldn’t accept it, in principle, the program works, but I would have enabled this option... If I turn it on, will you accept? It seems to me that the program will work more smoothly.

**Comment by ShvonderSiN** [comment]
Changes have been made, please re-consider. 
RIght, added the option - --device=dri so that there are no errors when displaying beautiful html pages

**Comment by hfiguiere** [comment]
> Changes have been made, please re-consider.

I can't you still haven't applied the other changes, include the source access that doesn't allow to even build.

**Comment by ShvonderSiN** [comment]
Changes have been made, please re-consider.
Oh my God, it's because I copied the wrong thing again) sorry for these ridiculous mistakes, this is the first time I've done this, I've returned github to the source again

File: org.mintguide.WebAppReader.metainfo.xml
Comment (hfiguiere) [inline org.mintguide.WebAppReader.metainfo.xml:1]: since this is upstream there is no reason to have that file here.

File: org.mintguide.WebAppReader.desktop
Comment (hfiguiere) [inline org.mintguide.WebAppReader.desktop:1]: this should be upstream. it's a bug to not have it.

File: org.mintguide.WebAppReader.yaml [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+      - type: file
+        path: org.mintguide.WebAppReader.desktop
+      - type: script
+        commands:
+          - 'python3 /app/webappreader/webappreader.py "$@"'
+        dest-filename: webappreader.sh
+      - type: file
+        path: org.mintguide.WebAppReader.metainfo.xml
+    post-install:
+      - install -Dm644 src/logo.png /app/share/icons/hicolor/128x128/apps/org.mintguide.WebAppReader.png
```
Comment (hfiguiere) [inline org.mintguide.WebAppReader.yaml:57]: the image is 80x80 which is:

1. too small
2. not 128x128

**Review by hfiguiere — CHANGES_REQUESTED** [review]
There are still changes you ignored.

File: org.mintguide.WebAppReader.desktop
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.desktop:1]: I don't understand where to store it? someone said above that this file is missing

Comment (hfiguiere) [reply org.mintguide.WebAppReader.desktop:1]: it should be along the application source code. For the last 25 years, desktop application expect to have a .desktop file. (nothing flatpak specific)

I said it was missing because a source entry was referring to it. And given that for the longest time the repository was 404 I couldn't check further.

File: org.mintguide.WebAppReader.yaml (hunk 10 above)
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:56]: I just don't understand what I have to do.

File: org.mintguide.WebAppReader.desktop
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.desktop:1]: so? now is ok? desktop file beside the manifest file

File: org.mintguide.WebAppReader.yaml (hunk 20 above)
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.yaml:57]: right, hope will find another sizes

File: org.mintguide.WebAppReader.yaml (hunk 10 above)
Comment (hfiguiere) [reply org.mintguide.WebAppReader.yaml:56]: > `post-install` should be merged `build-commands` since it's `simple` build system.

File: org.mintguide.WebAppReader.metainfo.xml
Comment (ShvonderSiN) [reply org.mintguide.WebAppReader.metainfo.xml:1]: should i delete metainfo file?

**Comment by ShvonderSiN** [comment]
I'm already very confused, very difficult to understand... sorry...
